### PR TITLE
Add LR scheduler to finetune/lora.py

### PR DESCRIPTION
**Pull request by rasbt**
For LLM training it's usually common practice to use a LR scheduler, which we haven't been doing in the past. What do you think about adding this as a hyperparameter option (defaulting to `False`).

What do you think? @carmocca @lantiga @awaelchli ?

 If you think that's a good idea, we can add it to all finetuning and pretraining scripts.

### CODE REVIEW

**Comment by awaelchli** [comment]
Not against it, but I'd like to suggest that if we add the scheduler, we enable it. With the minimalist philosophy of the project, we should aim at having as little code as possible deactivated, and make every line count. 

The finetuning scripts have learning rate warmup. What happens to it? Will you keep it? Furthermore, we would have to include the scheduler state in the checkpoint, so that we can reload it correctly.

**Comment by rasbt** [comment]
I wouldn't mind making it the default @awaelchli , and that's a good point regarding the warmup rate. We would have to do some experiments. Based on a single benchmark, the scheduler seems to be doing well (maybe the difference is negligible, but it doesn't cause any harm, and given that many people may wonder why we don't use a scheduler, this would be in favor of it):

```
+------------------+----------------+----------------+
|   Model Name     | truthfulqa_mc1 | truthfulqa_mc2 |
+------------------+----------------+----------------+
| Llama-2-7b-hf    |     0.2534     |     0.3967     | 
+------------------+----------------+----------------+
| Default LoRA     |     0.2815     |     0.4206     | 
+------------------+----------------+----------------+
| QLoRA (bnb.nf4)  |     0.2803     |     0.4139     | 
+------------------+----------------+----------------+
| QLoRA + scheduler|     0.2876     |     0.4216     | 
+------------------+----------------+----------------+
```

The above is based on 50k Alpaca iterations. It takes about 3 hours with QLoRA (each) so it's not a super exhaustive study.

I need to run an experiment without warmup though and see how that does.

**Comment by rasbt** [comment]
I enabled the scheduler by default according to your feedback @awaelchli 

I also made another small modification so that the scheduler is applied after the warmup step (previously, the scheduler and warmup were both used at the same time during warmup, which would be weird).

I also did an ablation study without warmup (last row). Here, I am using the learning rate that is used after the 100 default warmup steps as the initial learning rate:

```
+------------------+----------------+----------------+
|   Model Name     | truthfulqa_mc1 | truthfulqa_mc2 |
+------------------+----------------+----------------+
| Scheduler & warmup |     0.2876     |     0.4216     | 
+------------------+----------------+----------------+
| Scheduler after warmup |     0.2815     |     0.4228     | 
+------------------+----------------+----------------+
| Scheduler, no warmup|     0.2803     |     0.4147     | 
+------------------+----------------+----------------+
```

(Also note that mc2  is more important than mc1; mc2 is what everyone uses on the HF benchmarks afaik)

So, in practice there may be little difference but this is only on a single dataset and a single model. I would say we should keep the scheduler because it's common practice in LLM research. And for those who don't want it, it's probably easy to comment out the `scheduler.step()` line. (In contrast, adding it is slightly more complicated than removing it).

What do you think? If this sounds good then I can also apply that to the adapter scripts, pretraining, and full finetuning.

**Review by carmocca — COMMENTED** [review]
In addition to the suggestion of including warmup in the scheduler, we could also try the reciprocal-sqrt as proposed in https://twitter.com/__kolesnikov__/status/1687911223096971264/photo/1
